projects/igniteui-angular/src/lib/calendar/calendar-base.ts

```diff
@@ -651,54 +662,49 @@ export class IgxCalendarBaseDirective implements ControlValueAccessor {
 
             this.selectedDates = Array.from(new Set([...newDates, ...selDates])).map(v => new Date(v));
         } else {
-            const valueDateOnly = this.getDateOnly(value);
             let newSelection = [];
 
-            if (this.shiftKey && this._lastSelectedDate) {
+            if (this._shiftKey && this._lastSelectedDate) {
                 let start: Date;
                 let end: Date;
 
-                if (this._lastSelectedDate.getTime() < valueDateOnly.getTime()) {
-                    start = this._lastSelectedDate;
-                    end = valueDateOnly;
-                } else {
-                    start = valueDateOnly;
-                    end = this._lastSelectedDate;
-                }
+                [start, end] = this._lastSelectedDate.getTime() < value.getTime()
+                    ? [this._lastSelectedDate, value]
+                    : [value, this._lastSelectedDate];
 
-                // select all dates from last selected to shift clicked day
-                if (this.selectedDates.some((date: Date) => date.getTime() === this._lastSelectedDate.getTime())) {
+                const unselectedDates = [start, ...this.generateDateRange(start, end)]
+                    .filter(date => !this.isDateDisabled(date)
+                        && this.selectedDates.map(d => d.getTime()).indexOf(date.getTime()) === -1
```

Comment:
Not sure if filter + map is the best option here..need to look again. For sure === -1 can be replaced with using yhe `some` operator I think..you check it